[develop] Round 2 of overhaul to WE2E test suites (and other test improvements!)

mkavulich opened the pull request:
**Note to code manager: The label `run_we2e_fundamental_tests` should be renamed to `run_we2e_coverage_tests` after this PR is merged.**

## DESCRIPTION OF CHANGES: 
This test continues the overhaul of WE2E test suites as described in Issue #587 (specifically stage 3 and parts of stage 4 in [this comment](https://github.com/ufs-community/ufs-srweather-app/issues/587#issuecomment-1461049391)). The changes are summarized below, roughly in order of importance.

 - "fundamental" tests are replaced by "coverage" test suites. "fundamental" tests are returned to their original purpose: a lightweight set of tests to be run the same on all platforms. "coverage" tests now evenly distribute all comprehensive tests across all platform/compiler combinations for use in Jenkins testing.
 - "comprehensive" test list is updated to include all tests (except current known failures). For platforms that have known failures (for example, HPSS tests on on platforms without HPSS access), `comprehensive.<platform>[.<compiler>]` files are included to automatically run only the tests expected to succeed
 - Fix several existing failures
   - Use correct date format in grid_SUBCONUS_Ind_3km_ics_FV3GFS_lbcs_FV3GFS_suite_WoFS_v0
   - In `config_parser.py`, when populating a jinja template, keep dates in string format rather than converting to a `datetime` object (this fixes problem with `get_from_NOMADS_ics_FV3GFS_lbcs_FV3GFS`)
   - Fix unit tests for `retrieve_data.py`, there was a bug causing all tests to be run in nested subdirectories that eventually leads to failure when running all tests including HPSS retrieval
 - Remove several "get_from_HPSS" tests in favor of new unit tests for HPSS data in `test_retrieve_data.yaml`
 - Add several more dates and data sources to unit tests in `test_retrieve_data.yaml`
 - The example config files in the `ush/` directory (`config.community.yaml` and `config.nco.yaml`) are now included as WE2E tests (symbolically linked in the `tests/WE2E/test_configs/default_configs/` directory)
 - Remove long-known failing test `grid_RRFS_CONUS_25km_ics_GSMGFS_lbcs_GSMGFS_suite_GFS_v16` (#359). This is now an old capability with only legacy support (global spectral model was retired in 2019) and there are no immediate plans to fix the bug.
 - `WE2E_summary*.txt` files are now written to the experiment directory rather than `tests/WE2E`
 - Updated data_locations.yaml for latest RAP files on HPSS
 - Reduce timeouts and delays between calls to wget to speed up remote data retrieval
 - run_MET_GridStat_vx_APCP tasks fail randomly on occasion; increasing maxtries to 2 mitigates this problem
 - Swap test of restart capabilty from `grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16` to `grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v17_p8` for coverage reasons
 - Convert some `print_info_msg` messages to `logging.debug` calls to allow suppression of superfluous output if desired
 - A few miscellaneous minor fixes to log messages
 - Made some docstrings more consistent format
 - Removed some outdated documentation on validating config.yaml 

### Notes on current test limitations

1. Coverage tests will result in failures for Hera with intel due to outstanding known failures in NCO mode as described in issues #688 and #652 (as inherited from previous "fundamental" testing, the coverage tests for Hera with intel are all run in NCO mode). These should be run with caution until these issues are resolved.
2. The test `grid_RRFS_NA_3km_ics_FV3GFS_lbcs_FV3GFS_suite_RRFS_v1beta` is a known failure currently, and so has been removed from the pool of tests for now. This problem is described in issue #731
3. Currently some tests are failing on Orion due to a problem with the staged test data. I am working with @natalie-perlin to get this resolved.
4. Several tests have not yet completed as of the opening of this PR; will update the test section below as needed.

### Type of change
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## TESTS CONDUCTED: 

- [x] hera.intel
  - Had failures that were resolved on re-run
  - Ran `test_retrieve_data.py` successfully
- [x] hera.gnu
  - Coverage tests successful
- [x] orion.intel
  - Coverage tests successful
- [ ] cheyenne.intel
  - Tests in progress
- [x] cheyenne.gnu
  - Coverage tests successful
- [ ] gaea.intel
- [x] jet.intel
  - Coverage tests successful
- [ ] wcoss2.intel
- [ ] NOAA Cloud (indicate which platform)
- [ ] Jenkins
- [x] fundamental test suite
  - The "new" (restored old functionality) fundamental tests were run on Hera (intel) successfully
- [x] comprehensive tests
  - The new suite was run on Hera successfully

## DEPENDENCIES:
None.

## DOCUMENTATION:
Documentation for WE2E tests, including the table of test descriptions, has been updated. You can review the built documentation here: https://ufs-srweather-app-mkavulich.readthedocs.io/en/latest/WE2Etests.html

## ISSUE: 
 - Resolves #727
 - Resolves #705
 - Resolves #521
 - Resolves #347 
 - Resolves #742
 - Part of #587 though I will keep that issue open as there are still some things that need to be implemented (though I may consolidate the remaining work into a new issue for clarity).

## CHECKLIST
<!-- Add an X to check off a box. -->
- [x] My code follows the style guidelines in the Contributor's Guide
- [x] I have performed a self-review of my own code using the Code Reviewer's Guide
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] Any dependent changes have been merged and published

mkavulich committed:
Add example files 'config.community.yaml' and 'config.nco.yaml' from ush/ directory to list of valid tests via symlinks in new directory test_configs/default_configs/. Also include these new tests in the comprehensive suite.

mkavulich committed:
Write WE2E summary file to experiment directory instead of current directory, standardize function docstrings

mkavulich committed:
Remove references to validating config.yaml with ./config_utils.py; generate_FV3LAM_wflow.py should always output descriptive error messages if invalid config.yaml is provided

mkavulich committed:
- Convert several WE2E tests (retrieving data from HPSS) to unit tests
 - Fix chdir bug in test_retrieve_data.py
 - Relax timeout and delay times for wget commands in retrieve_data.py
 - Various minor code fixes

mkavulich committed:
Update latest RAP hpss filenames, add unit tests for RAP hpss dates

mkavulich committed:
- Cut down on intermediate test dates for retrieving data
 - Add test date for early RAP data with ICS
 - Retrieve RAP 09z out to 45 hours

mkavulich committed:
Still dealing with random failures for GridStat, bump up "maxtries" to 2

mkavulich committed:
Fix incorrect date format for grid_SUBCONUS_Ind_3km_ics_FV3GFS_lbcs_FV3GFS_suite_WoFS_v0 test

mkavulich committed:
Fix a few unit test errors for retrieve_data.py, all now pass

mkavulich committed:
For extend_yaml, when rendering a jinja template that returns a date-like object, keep it as a string. This fixes NOMADS test error (and any other test using the "days_ago" template)

mkavulich committed:
Don't overwrite existing walltime for re-run test

mkavulich committed:
Update comprehensive test suite to include all working tests, remove "wontfix" test grid_RRFS_CONUS_25km_ics_GSMGFS_lbcs_GSMGFS_suite_GFS_v16

mkavulich committed:
Rename "fundamental" tests to "coverage", revert "fundamental" test suite to its original purpose (small set of cheap tests to run on any machine)

mkavulich committed:
Skip any lines in read file that have comment character

mkavulich committed:
Swap restart test from grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16 to grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v17_p8

mkavulich committed:
Implement new, wider-coverage fundamental test suite. Still comes in under 100 core hours

mkavulich committed:
Add back NCO tests that are now working

mkavulich committed:
Remove grid_RRFS_NA_3km_ics_FV3GFS_lbcs_FV3GFS_suite_RRFS_v1beta test for now, it is not working; see issue 731

mkavulich committed:
Distribute all comprehensive tests among all platforms via "coverage" tests

mkavulich committed:
Increase grid_RRFS_NA_3km_ics_FV3GFS_lbcs_FV3GFS_suite_RRFS_v1beta run_post tasks to use 12 nodes to resolve #705

mkavulich committed:
Up run_fcst walltime to 2 hours for GST_release_public_v1 to prevent occasional walltime-related failures

mkavulich committed:
comprehensive file for Cheyenne (omitting all "get_from" tests since compute nodes have no internet connection)

mkavulich committed:
Shuffle some tests around to avoid running "get_from" tests on Cheyenne

mkavulich committed:
Put WE2E_summary file in correct directory (level above experiment directories)

mkavulich committed:
Shuffle coverage tests a bit more, reduce duplication of 3km tests, remove expensive tests from GNU platforms

mkavulich committed:
Convert print_info_msg "VERBOSE" statements to logging.debug to allow suppression of unnecessary output at generate step

mkavulich committed:
Add logic to skip new HPSS unit test on non-HPSS platforms

MichaelLueken reviewed (approved):
@mkavulich These changes look good to me!  Will submit Jenkins tests to make sure that the switch from fundamental to coverage works properly.

mkavulich commented:
Thanks @MichaelLueken. Natalie got back to me recently and let me know that Orion should be working now, so hopefully those tests will pass!

Note that I think I made the necessary updates to the CI/CD scripts so that they will run the new "coverage" test suite, but I have not actually tested that interface yet. I will check on the results of those tests to make sure the correct sets of tests have been run.

mkavulich committed:
Alphabetize coverage test files

MichaelLueken commented:
@mkavulich Looking through the results of the Jenkins tests, I have noted failures:

During the taring stage after the tests ran, the `tests/WE2E/WE2E_summary*txt` file is no longer present.  This file is currently required to be present.  Since it is missing, the tar stage fails.  Please redirect the path in `.cicd/Jenkinsfile` from `tests/WE2E` to `expt_dirs`

On Cheyenne Intel, three tests failed:
```
grid_RRFS_CONUS_25km_ics_NAM_lbcs_NAM_suite_GFS_v16
grid_RRFS_CONUS_3km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15p2
nco_grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_timeoffset_suite_GFS_v16
```

I believe that these tests failed due to naming convention of the directory - `/glade/scratch/epicufsrt/jenkins/workspace/fs-srweather-app_pipeline_PR-732@2` and the way that rocoto handles the `@` symbol.  The EPIC platform team has changed the `@` to a `_`, so this should no longer be an issue moving forward.  I will attempt to rerun the tests following this change to see if they continue to fail.

On Gaea, one test failed:

`get_from_NOMADS_ics_FV3GFS_lbcs_FV3GFS`

Unfortunately, Gaea is down today, making it difficult to see what caused the issue.

On Hera GNU, three tests failed:
```
get_from_HPSS_ics_FV3GFS_lbcs_FV3GFS_fmt_nemsio_2019061200
get_from_HPSS_ics_GSMGFS_lbcs_GSMGFS
nco_grid_RRFS_CONUS_13km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16
```

The directory of these tests is - `/scratch1/NCEPDEV/stmp2/role.epic/jenkins/workspace/fs-srweather-app_pipeline_PR-732/expt_dirs`.  Looking in the logs, there appears to be an issue with the keytab auth method used by the role.epic account to access HPSS.  I will need to reach out to the EPIC platform team to have this addressed.

On Hera Intel, all tests failed.  Similar to Cheyenne Intel, this is likely due to the naming convention of the concurrent build on Hera - `/scratch1/NCEPDEV/stmp2/role.epic/jenkins/workspace/fs-srweather-app_pipeline_PR-732@2`.  This should have been addressed by the platform team, so I can attempt to rerun once you have given the go ahead.

On Jet, four tests failed:
```
community
get_from_HPSS_ics_FV3GFS_lbcs_FV3GFS_fmt_nemsio_2021032018
get_from_HPSS_ics_FV3GFS_lbcs_FV3GFS_fmt_netcdf_2022060112_48h
get_from_HPSS_ics_RAP_lbcs_RAP
```

These tests can be found - `/mnt/lfs4/HFIP/hfv3gfs/role.epic/jenkins/workspace/fs-srweather-app_pipeline_PR-732` and the keytab auth method is once again the culprit for these failures.

All tests successfully passed for both Cheyenne GNU and Orion.

danielabdi-noaa commented on .cicd/scripts/srw_test.sh:
```diff
@@ -34,7 +34,7 @@ nco_dir="${workspace}/nco_dirs"
     test_type="comprehensive"
 else
-    test_type="fundamental"
+    test_type="coverage"
```
The corresponding label "run_we2e_fundamental_tests" should be changed accordingly.

mkavulich replied:
Good catch, I think this should happen after the PR is merged though.

danielabdi-noaa commented on ush/test_retrieve_data.py:
```diff
@@ -30,78 +34,130 @@ class FunctionalTesting(unittest.TestCase):
         """In real time, GDAS is used for LBCS with a 6 hour offset."""
 
         with tempfile.TemporaryDirectory(dir=".") as tmp_dir:
-            os.chdir(tmp_dir)
```
I don't think this should be remove. In NCO mode the locations were conflicting, so I had to use temp directory for it.
This is the commit that changed it
https://github.com/danielabdi-noaa/ufs-srweather-app/commit/f3b1b3f75a051a07d06f59f6caccfd8f4e277705

mkavulich replied:
Sorry I don't understand, can you clarify what NCO mode means specifically for these tests? Shouldn't the unit tests be run the same way regardless of NCO or community mode? Here the tmpdir is still used as the destination of the test data, we just don't use that as a working directory.

If the "os.chdir" lines are restored, we will need to include a "chdir .." at the end to avoid this problem of recursively nesting tmpdirs.

danielabdi-noaa replied:
Cleanup was not working for NCO mode at the time, and `clean_up_output_dir` still uses `os.getcwd()`. The current path should be EXPTDIR/CDATE for community mode and COMIN for NCO mode for it to work properly. The ex-scripts that call `retrieve_data.py` switch to those directories first, so `test_retrieve_data.py` should do the same and do `chdir ..` later to fix the recursive issue you pointed out.

mkavulich replied:
@danielabdi-noaa I have restored the usage of the tmpdir in the unit tests (using an absolute path to avoid previous issues). I will point out that this is not a requirement on the retrieve_data.py script, so theoretically that's something that should be covered in a test of the exscripts, not this unit test. But we can keep it here while we do not have a test specific to the exscripts.

danielabdi-noaa commented:
@MichaelLueken The "@" issue was solved in the old xml file but I think we reverted back to enclosing the whole LOGDIR with <cyclestr> after the latest workflow changes, so only a server side solution will work now.

MichaelLueken commented:
@danielabdi-noaa It looks like during the migration, since Jesse is no longer with EPIC, the platform team just started up the runners without having prior knowledge that the default `@` doesn't play will with rocoto.  My current tests are showing that platform team have gone through and successfully changed these, so we are once again back to `__` rather than `@`.

mkavulich committed:
Update Jenkinsfile for new summary file location

mkavulich commented:
@MichaelLueken Thanks for the summary. If I'm reading this correctly it looks like the only action needed from me for the test failures is to fix the location of the summary file in the Jenkinsfile, correct? Keep me posted if you need anything else on my end.

MichaelLueken commented:
@mkavulich Correct, the majority of the failures encountered were due to issues that the EPIC platform team will need to address (PR #737 has shown that the problem encountered due to path names should no longer exist).  Once I receive feedback about the new helpdesk issue I opened this morning about HPSS on Hera and Jet, I should be able to move forward with rerunning this PR.

mkavulich committed:
Omit new "default_configs" tests from WE2E_test_info.txt, due to format limitations

mkavulich committed:
Updates to documentation

 - Add glossary entries for "J-jobs" and "exscripts"
 - Update WE2Etests.rst for new testing suites, test directories
 - Update examples for latest code

mkavulich committed:
Remove "Creating Alternate Names for a Test" section; this is several paragraphs describing the creation of a symbolic link, something which is already described in a few sentences above

mkavulich committed:
Update table of WE2E tests

mkavulich committed:
Add core hours and walltime info to table

MichaelLueken commented:
@mkavulich Looking at the Gaea test that failed - `get_from_NOMADS_ics_FV3GFS_lbcs_FV3GFS`, it looks like the test is attempting to pull data from HPSS rather than NOMADS.  Looking at the logs, it should be pulling from NOMADS, so I'm not sure why the test is trying to pull from HPSS.  The test can be found - `/lustre/f2/dev/wpo/role.epic/jenkins/workspace/fs-srweather-app_pipeline_PR-732/expt_dirs/get_from_NOMADS_ics_FV3GFS_lbcs_FV3GFS`.

mkavulich committed:
For NOMADS test, EXTRN_MDL_DATA_STORES should be set explicitly to "nomads"

mkavulich committed:
Add new orion comprehensive file; HPSS not available, but AWS and NOMADS tests should work

mkavulich committed:
Restore "chdir" to tmpdir in test_retrieve_data.py per PR review comments

mkavulich commented:
@MichaelLueken thanks for keeping me updated. I have pushed a fix to the NOMADS test (not sure why it was including checks for HPSS and AWS as well).

Note that I am not sure if the NOMADS test *should* work on Gaea (since I can not test it myself), so if it fails again I can move that test to another machine in the coverage set.

mkavulich committed:
Fix date error in FV3GFS netcdf test, add pylint suggestions (now lints 10/10)

MichaelLueken commented:
@mkavulich The majority of the Jenkins issues have been addressed.  The EPIC role account has also been granted access to HPSS, but we don't currently have rstprod access.  This is causing the `get_from_HPSS_ics_GSMGFS_lbcs_GSMGFS` test to fail in the `get_extrn_ics` task.  I've reached out to the RDHPCS help desk to see if it is possible to add rstprod access to a role account via AIM.  If it is, then the EPIC program office will put in the request.  My latest run of your PR is using a sandbox pipeline so that your modified Jenkinsfile will be used.  If you would like to check on the progress on Hera, the path is `/scratch1/NCEPDEV/stmp2/role.epic/jenkins/workspace/line_-_Remote_Jenkinsfile_PR-732`.

On other machines, replacing `fs-srweather-app_pipeline_PR-732` with `line_-_Remote_Jenkinsfile_PR-732` will allow you to check the experiments.

mkavulich committed:
Load wget module on Orion; this is a suggestion from Orion support to fix errors in wget due to the built-in wget on Orion being quite old

mkavulich commented:
@MichaelLueken Thanks for kicking off the tests again. Note that there may be some failures on Orion still: I discovered some more static data directories that do not have read permissions. Once that is fixed the tests should succeed.

I also added "wget" to the list of modules to load on Orion; this was a suggestion from the Orion helpdesk to solve some wget errors for the NOMADs test. This shouldn't affect the results since the NOMADS test is not run on Orion for the coverage tests, but I thought I would mention it in case other failures pop up.

MichaelLueken commented:
@mkavulich The GSMGFS test that is failing due to attempting to pull restricted data is `get_from_HPSS_ics_GSMGFS_lbcs_GSMGFS` in the Hera GNU coverage suite.

On Gaea, `get_from_NOMADS_ics_FV3GFS_lbcs_FV3GFS` is still failing.  The reason for the failure is:

`/bin/sh: wget: command not found`

It might be better to replace the `get_from_HPSS_ics_GSMGFS_lbcs_GSMGFS` test with `get_from_NOMADS_ics_FV3GFS_lbcs_FV3GFS`

mkavulich committed:
Move NOMADS test from Gaea to Hera GNU due to unavailability of that capability on Gaea; remove get_from_HPSS_ics_GSMGFS_lbcs_GSMGFS test due to data restrictions...developers have decided that this legacy capability is not worth fretting over

mkavulich committed:
Fix plotting task workflow yaml to run correct script

mkavulich commented:
@MichaelLueken The changes are now in for running different tests (I also included a fix for a bug in the plotting script in #742); unless there is a failure I missed the only coverage tests that should need re-running prior to merge are Gaea and Hera/GNU.

MichaelLueken commented:
@mkavulich PR #736 was merged earlier this morning.  This PR updated the test_retrieve_data.py unittest script.  Please merge the latest develop into your branch to correct the conflict.  Thanks!

MichaelLueken commented:
@mkavulich The latest testing has completed.  All tests pass on both Gaea and Hera GNU.  A rerun of Cheyenne Intel, however, is still showing two persistent failures (these failures were previously noted, but the cause of the error was blamed on directory naming).

The `grid_RRFS_CONUS_3km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15p2` test is failing in the `run_post_mem000_f00*` tasks with the following message:

`This post cannot produce IFI icing products because it was not compiled with libIFI.`

The Jenkins directory for this experiment is - `/glade/scratch/epicufsrt/jenkins/workspace/line_-_Remote_Jenkinsfile_PR-732__2/expt_dirs/grid_RRFS_CONUS_3km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15p2`.

The `nco_grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_timeoffset_suite_GFS_v16` test is failing in `get_extrn_ics` with the following message:

```
INFO: Getting file: /glade/p/ral/jntp/UFS_CAM/COMGFS/gfs.20220810/06/gfs.t06z.sfcf006.nc

INFO: File does not exist on disk
 /glade/p/ral/jntp/UFS_CAM/COMGFS/gfs.20220810/06/gfs.t06z.sfcf006.nc
 try using: --input_file_path <your_path>
```
The Jenkins directory for this experiment is - `/glade/scratch/epicufsrt/jenkins/workspace/line_-_Remote_Jenkinsfile_PR-732__2/expt_dirs/nco_grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_timeoffset_suite_GFS_v16`.

I can see that there is no `gfs.t06z.sfcf006.nc` file in the expected directory.  Do we need to reach out to the AUS team to see why this file isn't present on Cheyenne?

MichaelLueken commented:
@mkavulich It looks like moving the `grid_RRFS_CONUS_3km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15p2` test to Hera should work (Hera includes the new IFI library).  I'll test it there and see how it works.

mkavulich committed:
Merge remote-tracking branch 'origin/develop' into feature/WE2E_test_improvement_round_2

mkavulich committed:
Temporarily remove nco_grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_timeoffset_suite_GFS_v16 test, data needs to be staged

mkavulich committed:
Move grid_RRFS_CONUS_3km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15p2 test to Hera, intel due to bad libraries problem on Cheyenne

mkavulich commented:
@MichaelLueken I have merged in the latest changes and made changes to the test files so the coverage tests should now succeed. 

I can not figure out why the test_retrieve_data unit tests are failing. They are running successfully (albeit *very* slowly due to the large file size) on Hera. Do you have any insight on this? Without being able to replicate it I don't know how to solve this problem.

MichaelLueken commented:
@mkavulich Interestingly, while attempting to run the failing unit test on Hera using your branch, I'm seeing the same failures that I have noted below regarding the UFS-CASE-STUDY ICs and LBCs from AWS.

Looking at the details for the failed `Python functional tests`, I'm seeing the following:

```
ERROR: test_ufs_ics_from_aws (test_retrieve_data.FunctionalTesting)
Get UFS-CASE-STUDY ICS from aws
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/ufs-srweather-app/ufs-srweather-app/ush/test_retrieve_data.py", line 459, in test_ufs_ics_from_aws
    retrieve_data.main(args)
  File "/home/runner/work/ufs-srweather-app/ufs-srweather-app/ush/retrieve_data.py", line 833, in main
    unavailable = get_requested_files(
  File "/home/runner/work/ufs-srweather-app/ufs-srweather-app/ush/retrieve_data.py", line [36](https://github.com/ufs-community/ufs-srweather-app/actions/runs/4805465573/jobs/8552121179?pr=732#step:4:37)1, in get_requested_files
    orig_path = os.getcwd()
FileNotFoundError: [Errno 2] No such file or directory
```
and
```
ERROR: test_ufs_lbcs_from_aws (test_retrieve_data.FunctionalTesting)
Get UFS-CASE-STUDY LBCS from aws for 3 hour boundary conditions
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/ufs-srweather-app/ufs-srweather-app/ush/test_retrieve_data.py", line 488, in test_ufs_lbcs_from_aws
    retrieve_data.main(args)
  File "/home/runner/work/ufs-srweather-app/ufs-srweather-app/ush/retrieve_data.py", line [83](https://github.com/ufs-community/ufs-srweather-app/actions/runs/4805465573/jobs/8552121179?pr=732#step:4:84)3, in main
    unavailable = get_requested_files(
  File "/home/runner/work/ufs-srweather-app/ufs-srweather-app/ush/retrieve_data.py", line 361, in get_requested_files
    orig_path = os.getcwd()
FileNotFoundError: [Errno 2] No such file or directory
```
It appears as though the unit test is failing due to the inability to find the the files that are supposed to be pulled from AWS.  Looking at the file naming convention for the CAPE 2020 case study, the data being pulled should be correct:

`INFO: Getting files named like ['{yyyymmddhh}.gfs.nemsio.tar.gz']` for ICs and `INFO: Getting files named like ['{yyyymmddhh}_bc.atmf{fcst_hr:03d}.nemsio.tar.gz']` for LBCs.

Unfortunately, the DEBUG level prints aren't available in the log for the failed unit test, so there are no messages like:

```
DEBUG: Looking for fhr = 6 
 
DEBUG: Looking for files like ['gep{mem:02d}.t{hh}z.pgrb2a.0p50.f{fcst_hr:03d}', 'gep{mem:02d}.t{hh}z.pgrb2b.0p50.f{fcst_hr:03d}'] 
 
DEBUG: They should be here: ['https://noaa-gefs-pds.s3.amazonaws.com/gefs.{yyyymmdd}/{hh}/atmos/pgrb2ap5', 'https://noaa-gefs-pds.s3.amazonaws.com/gefs.{yyyymmdd}/{hh}/atmos/pgrb2bp5']
```
At this point, I can only think that the files that the unit test is attempting to pull doesn't include `https://ufs-case-studies.s3.amazonaws.com/` in the URL, causing the test to fail to find the data.

mkavulich committed:
Trying to solve server-side CI problems errors (local tests are passing)

MichaelLueken commented:
@mkavulich The unit tests are now passing.  I will retest Cheyenne Intel and Hera Intel, then merge this work.  Thank you very much!

mkavulich commented:
Okay, I think I figured out what was happening (thought I still don't understand it). The new tests didn't have the "chdir" commands requested by @danielabdi-noaa, and for some reason those tests specifically were failing without the chdir commands (despite others working). The problem didn't appear for me because I was running those tests individually rather than as part of the full set of tests. I pushed a change to include those lines and it seems to be working now. 

Regarding the older failures, the `nco_grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_timeoffset_suite_GFS_v16` is a weird one that I had not noticed before: it is meant to be using staged netCDF data from FV3GFS, but that data isn't staged in the Epic locations. Jet and Hera tests are only succeeding due to the fallback to HPSS retrieval, and on Cheyenne it was still pointing to a legacy DTC space staged location (which was still missing files). I have added those files to the DTC space so it should pass now, but I've removed it from the list so that it can be re-added later with properly staged data.

I have merged in the latest changes and swapped the `grid_RRFS_CONUS_3km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15p2` test to Hera, so *hopefully* everything should be passing now.

MichaelLueken commented:
The unit tests have successfully passed and reruns of Cheyenne Intel and Hera Intel show that the tests pass (with the exceptions of verification tests, outlined in issue #688).  Merging this PR to develop now.

MichaelLueken commented:
Renamed the `run_we2e_fundamental_tests` label to `run_we2e_coverage_tests`.  Updated the Jenkins automated testing pipeline to pick up PRs with the new `run_we2e_coverage_tests` label rather than `run_we2e_fundamental_tests`.
